### DISCUSSION TITLE
get_user_db dependency

### DISCUSSION THREAD

**Discussion by davidbrochart** [fastapi-users/fastapi-users#736] [Q&A]
I understand `get_user_db` is a way to abstract out the DB and use it as a dependency callable. But there are other (non-router) places where it might be needed, and where it would be nice to e.g. mock it for testing, such as:
- `startup`/`shutdown` event handlers,
- `CookieAuthentication` subclasses.

I guess we can't use a dependency on these?

**Comment by frankie567** [comment]
Indeed, but this was actually already the case before where we would instantiate the user database class independently. Actually, nothing prevents you to do it manually; the only drawback is that you won't be able to override them through the FastAPI dependency system. In this case, you could patch the methods using a magic mock.

```py
async def create_token_user():
    # ...
    TOKEN_USER = UserDB(id=USER_TOKEN, email=token_email, hashed_password="")
    user_db = SQLAlchemyUserDatabase(UserDB, database, users, oauth_accounts)
    await user_db.create(TOKEN_USER)
```

You can do the same inside a `get_login_response` if needed.

**Reply by frankie567** [reply]
That said, maybe it could worth it to pass `user_manager` as a parameter of `get_login_response` and `get_logout_response`. It could ease the implementation of database-based access tokens. I'll think about it!

**Reply by frankie567** [reply]
So, I was quite seduced by the idea so I just released it in [`v8.1.0`](https://github.com/fastapi-users/fastapi-users/releases/tag/v8.1.0) 😄

So basically, it could look like this:

```py
class LoginCookieAuthentication(CookieAuthentication):
    async def get_login_response(self, user, response, user_manager):
        await super().get_login_response(user, response, user_manager)
        # set user as logged in
        user.logged_in = True
        await user_manager.user_db.update(user)
        # auto redirect
        response.status_code = status.HTTP_302_FOUND
        response.headers["Location"] = "/lab"

    async def get_logout_response(self, user, response, user_manager):
        await super().get_logout_response(user, response, user_manager)
        # set user as logged out
        user.logged_in = False
        await user_manager.user_db.update(user)
```

It's working inside the dependency system, so overriding will 100% work in this context.

**Reply by davidbrochart** [reply]
Awesome, thanks a lot @frankie567 !
